sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/PushNotificationClient.java

```diff
@@ -185,4 +210,77 @@ private ChatEvent parsePushNotificationEvent(ChatEventType chatEventType, ChatPu
         this.logger.info(" Result: " + jsonString);
         return NotificationUtils.toEventPayload(chatEventType, jsonString);
     }
+
+    private void scheduleRegistrationRenew(final long delayMs, final int tryCount) {
+        if (!this.isPushNotificationsStarted) {
+            this.logger.info("Push notification has already been stopped! No need to renew registration.");
+            return;
+        }
+
+        if (this.deviceRegistrationToken == null) {
+            stopPushNotifications();
+            Throwable throwable = new Throwable("No device registration token stored!");
+            this.logger.logThrowableAsError(throwable);
+            if (this.registrationErrorHandler != null) {
+                this.registrationErrorHandler.accept(throwable);
+            }
+            return;
+        }
+
+        if (tryCount > NotificationUtils.MAX_REGISTRATION_RETRY_COUNT) {
+            stopPushNotifications();
+            Throwable throwable = new Throwable(
+                "Registration renew request failed after "
+                + NotificationUtils.MAX_REGISTRATION_RETRY_COUNT
+                + "retries.");
+            this.logger.logThrowableAsError(throwable);
+            if (this.registrationErrorHandler != null) {
+                this.registrationErrorHandler.accept(throwable);
+            }
+            return;
+        }
+
+        this.logger.info("Scheduling Registrar registration automatically renew in " + delayMs + "ms");
+
+        TimerTask task = new TimerTask() {
+            @Override
+            public void run() {
+                boolean retry = false;
+
+                PushNotificationClient.this.logger.info("Renew Registrar registration attempt #" + tryCount);
+
+                String skypeUserToken;
+                try {
+                    skypeUserToken = communicationTokenCredential.getToken().get().getToken();
```

Comment:
this chain of calls looks weird. Is this how it is ?

Reply:
Yes... 
The first getToken would return a CompletableFuture<CommunicationAccessToken>.
Then, second get() returns a CommunicationAccessToken.
Finally, the last getToken is CommunicationAccessToken.getToken() and would return a string.